### PR TITLE
saving words to redis

### PR DESCRIPTION
**Pull request by jjang4001**
Concurrently saving words to redis.

To see changes:
1. run  ```go run producer/producer.go```
2. run ```go run consumer/main.go```
3. run
```
redis-cli
GET <any word from tweet that was printed to console>
```
And you should get that the value is greater than or equal to 1.

INCR command sets it to 1 if the key doesn't exist btw.

### CODE REVIEW

File: datastore/redis.go [hunk 1]
```diff
@@ -55,6 +55,14 @@ func (r *RedisDatastore) GetExample(exampleId string) (*models.Example, error) {
 	return &models.Example{ExampleId: exampleId, ExampleValue: exampleVal}, nil
 }
 
+func (r *RedisDatastore) SaveWord(word string) error {
+	if r.Cmd("INCR", word).Err != nil {
```
Comment (philipslan) [inline datastore/redis.go]: put this in the common library as a constant

File: consumer/consumer/consumer.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+		return
+	}
+
+	parsedWords := make(chan string)
+	go parser.GetWordsFromTweet(tweet, parsedWords)
+	saveWord(parsedWords, consumer.DbEnv)
+	delivery.Ack()
+}
+
+func saveWord(parsedWords <-chan string, dbEnv *datastore.Env) {
```
Comment (philipslan) [inline consumer/consumer/consumer.go:45]: can you make the increments one atomic action? If so, we should throw if it ever fails and reject the delivery. If not, are we okay with part of the list not being incremented?

Comment (jjang4001) [reply consumer/consumer/consumer.go:45]: What do you mean one atomic action? Do you mean increment all words of the tweet at the same time?

File: consumer/consumer/consumer.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+	if err := json.Unmarshal([]byte(payload), &tweet); err != nil {
+		// handle error
+		delivery.Reject()
+		return
+	}
+
+	parsedWords := make(chan string)
+	go parser.GetWordsFromTweet(tweet, parsedWords)
+	saveWord(parsedWords, consumer.DbEnv)
```
Comment (philipslan) [inline consumer/consumer/consumer.go:37]: see comment below, should we have case where we reject the delivery?

File: consumer/main.go [hunk 4]
```diff
@@ -49,7 +27,14 @@ func main() {
...
-	taskConsumer := &taskConsumer{}
+	db, err := datastore.NewDatastore(datastore.REDIS, dbConnectionString)
+	if err != nil {
+		log.Print(err)
+	}
+	defer db.Close()
+
+	env := datastore.Env{DB: db}
+	taskConsumer := &consumer.TaskConsumer{DbEnv: &env}
 	taskQueue.AddConsumer("task consumer", taskConsumer)
```
Comment (philipslan) [inline consumer/main.go]: i missed this can you move this to the common constants file and rename it to `tweet consumer`

File: parser/parser_test.go [hunk 5]
```diff
@@ -1,15 +0,0 @@
...
-
-	"github.com/stretchr/testify/assert"
-)
-
-func Test_ParseString(t *testing.T) {
-	assert := assert.New(t)
-	input := "A little more than 24 hours from now Trump will declare victory.\n\nIf Republicans do well, Trump will take credit\n\nIf Democrats do well Trump will say it would have been much worse without his rallies. He will also blame “illegal” voters and the media.\n\nBookmark this."
-	output := PrintAndParse(input)
-	expected := []string{"little", "hours", "trump", "declare", "victory", "republicans", "well", "trump", "take", "credit", "democrats", "well", "trump", "say", "would", "much", "worse", "without", "rallies", "also", "blame", "illegal", "voters", "media", "bookmark"}
-	assert.Equal(expected, output)
```
Comment (philipslan) [inline parser/parser_test.go:14]: why is this file deleted if it still able to run and pass we should leave it to maintain the functionality of the underlying logic

File: parser/tests/PrintAndParse_test.go [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package tests
+
+import (
+	"testing"
+	"twittertracker/parser"
+)
+
+func Test_ParseString(t *testing.T) {
+	// assert := assert.New(t)
```
Comment (philipslan) [inline parser/tests/PrintAndParse_test.go]: why is this commented out?

File: consumer/consumer/consumer.go (hunk 2 above)
Comment (philipslan) [reply consumer/consumer/consumer.go:45]: yeah try to increment all or fail. If it fails that's fine because it just means that it'll be reprocessed. So ideally we want this to be an atomic event.

File: parser/tests/PrintAndParse_test.go (hunk 6 above)
Comment (jjang4001) [reply parser/tests/PrintAndParse_test.go]: This test became useless... I made it so that we save to db at the same time as parsing.. Which is why the atomically saving all the db will be difficult haha

File: consumer/consumer/consumer.go (hunk 2 above)
Comment (jjang4001) [reply consumer/consumer/consumer.go:45]: Why would running increment through all the words of the array at the end make it not fail? versus incrementing as soon as a word is parsed? won't the failure on a word be the same? 

Unless you mean that it's because of just processing too many things and it'll just have some random uncaught error that we didn't account for. But in this case, we don't even yet have a way of reprocessing on failure.

Comment (philipslan) [reply consumer/consumer/consumer.go:45]: it could be a failure in the machine or the db fails to increment. Any number of factors, but since we're trying the whole tweet as an entry we should treat adding it's data as one operation as well. If the payload fails it automatically gets reprocessed because the payload doesn't get marked as done in the queue. At least that should be how the message queue works. We can do some testing to confirm this.

Comment (philipslan) [reply consumer/consumer/consumer.go:45]: https://redis.io/topics/transactions

File: parser/parser_test.go (hunk 5 above)
Comment (jjang4001) [reply parser/parser_test.go:14]: It wasn't deleted. It was simply moved to PrintAndParse_test in the tests directory. Also, it is unable to pass since we no longer return an array, and PrintAndParse does not control anything with the channel. Literally the only thing that this would be testing is that the line
```
strings.Replace(text, "\n", " ", -1)
```
is doing what it's supposed to be doing.

Tbh, probs should've deleted this method and just call getWords instead inside of retrieveCorrectTypes.

Comment (philipslan) [reply parser/parser_test.go:14]: Yeah let’s do that get rid of print and parse my bad

File: parser/tests/PrintAndParse_test.go (hunk 6 above)
Comment (philipslan) [reply parser/tests/PrintAndParse_test.go]: We should still maintain this unit test. If you don’t want to I can fix it in another pr.

**Review by philipslan — APPROVED** [review]
